src/StructuredYson/StructuredYson.tsx

```diff
@@ -211,23 +233,62 @@ export class StructuredYson extends React.PureComponent<Props, State> {
         });
     };
 
+    findCollapsedParent = (matchPath: string, collapsedState: CollapsedState): string | null => {
+        const pathParts = matchPath.split('/');
+        for (let i = 1; i <= pathParts.length; i++) {
+            const checkPath = pathParts.slice(0, i).join('/');
+            if (collapsedState[checkPath]) {
+                return checkPath;
+            }
+        }
+        return null;
+    };
+
     onNextMatch = (_event: unknown, diff = 1) => {
-        const {matchIndex, matchedRows} = this.state;
-        if (isEmpty_(matchedRows)) {
+        const {matchIndex, matchedRows, allMatchPaths, collapsedState} = this.state;
+
+        const totalMatches = allMatchPaths?.length || 0;
+
+        if (totalMatches === 0) {
             return;
         }
 
-        let index = (matchIndex + diff) % matchedRows.length;
-        if (index < 0) {
-            index = matchedRows.length + index;
-        }
+        // Calculate next index in total matches
+        let nextTotalIndex = matchIndex + diff;
+        nextTotalIndex = ((nextTotalIndex % totalMatches) + totalMatches) % totalMatches;
+
+        const targetMatchPath = allMatchPaths[nextTotalIndex];
+
+        const collapsedParent = this.findCollapsedParent(targetMatchPath, collapsedState);
+
+        // If target is not hidden, it's visible - navigate to it
+        if (collapsedParent === null) {
+            // Count how many visible matches are before our target
+            let visibleMatchCount = 0;
+            for (let i = 0; i < nextTotalIndex; i++) {
+                if (this.findCollapsedParent(allMatchPaths[i], collapsedState) === null) {
+                    visibleMatchCount++;
+                }
+            }
 
-        if (index !== matchIndex) {
-            this.setState({matchIndex: index});
+            // Navigate to the visible match
+            if (visibleMatchCount < matchedRows.length) {
+                this.setState({matchIndex: nextTotalIndex});
+                this.tableRef.current?.scrollToIndex(matchedRows[visibleMatchCount]);
+                this.searchRef.current?.focus();
+            }
+            return;
         }
 
-        this.tableRef.current?.scrollToIndex(matchedRows[index]);
-        this.searchRef.current?.focus();
+        // Target is hidden - expand the collapsed parent
+        const newCollapsedState = {...collapsedState};
+        delete newCollapsedState[collapsedParent];
```

Comment:
As I understand it expands only one level for each call of `this.onNextMatch`. Can we expand all parents here before call retry `this.onNextMatch(...)`?